api/bucketsd/service.go

```diff
@@ -1497,6 +1518,367 @@ func (s *Service) PushPath(server pb.APIService_PushPathServer) (err error) {
+	}()
+
+	return fa, nil
+}
+
+func (s *Service) PushPaths(server pb.APIService_PushPathsServer) error {
+	log.Debugf("received push paths request")
+
+	ctx, cancel := context.WithCancel(server.Context())
+	defer cancel()
+
+	dbID, ok := common.ThreadIDFromContext(ctx)
+	if !ok {
+		return errDBRequired
+	}
+	dbToken, _ := thread.TokenFromContext(ctx)
+
+	req, err := server.Recv()
+	if err != nil {
+		return fmt.Errorf("on receive: %v", err)
+	}
+	var buckKey, buckRoot string
+	switch payload := req.Payload.(type) {
+	case *pb.PushPathsRequest_Header_:
+		buckKey = payload.Header.Key
+		buckRoot = payload.Header.Root
+	default:
+		return fmt.Errorf("push bucket path header is required")
+	}
+
+	lck := s.Semaphores.Get(buckLock(buckKey))
+	lck.Acquire()
+	defer lck.Release()
+
+	buck := &tdb.Bucket{}
+	err = s.Buckets.GetSafe(ctx, dbID, buckKey, buck, tdb.WithToken(dbToken))
+	if err != nil {
+		return fmt.Errorf("getting bucket: %v", err)
+	}
+	if buckRoot != "" && buckRoot != buck.Path {
+		return status.Error(codes.FailedPrecondition, buckets.ErrNonFastForward.Error())
+	}
+
+	var wg sync.WaitGroup
+	addedCh := make(chan addedFile)
+	doneCh := make(chan struct{})
+	errCh := make(chan error)
+	go func() {
+		queue := newFileQueue()
+		for {
+			req, err := server.Recv()
+			if err == io.EOF {
+				wg.Wait() // Request ended normally, wait for pending jobs
+				close(doneCh)
+				return
+			} else if err != nil {
+				errCh <- fmt.Errorf("on receive: %v", err)
+				return
+			}
+			switch payload := req.Payload.(type) {
+			case *pb.PushPathsRequest_Chunk_:
+				pth, err := parsePath(payload.Chunk.Path)
+				if err != nil {
+					errCh <- fmt.Errorf("parsing path: %v", err)
+					return
+				}
+				fa, err := queue.add(ctx, s.IPFSClient.Unixfs(), pth, func() ([]byte, error) {
+					buck.UpdatedAt = time.Now().UnixNano()
+					buck.SetMetadataAtPath(pth, tdb.Metadata{
+						UpdatedAt: buck.UpdatedAt,
+					})
+					buck.UnsetMetadataWithPrefix(pth + "/")
+					if err = s.Buckets.Verify(ctx, dbID, buck, tdb.WithToken(dbToken)); err != nil {
+						return nil, fmt.Errorf("verifying bucket update: %v", err)
+					}
+					key, err := buck.GetFileEncryptionKeyForPath(pth)
+					if err != nil {
+						return nil, fmt.Errorf("getting bucket key: %v", err)
+					}
+					wg.Add(1)
+					return key, nil
+				}, addedCh, errCh)
+				if err != nil {
+					errCh <- fmt.Errorf("enqueueing file: %v", err)
+					return
+				}
+
+				if len(payload.Chunk.Data) > 0 {
+					if _, err := fa.writer.Write(payload.Chunk.Data); err != nil {
+						errCh <- fmt.Errorf("writing chunk: %v", err)
+						return
+					}
+				} else {
+					if err := fa.writer.Close(); err != nil {
+						errCh <- fmt.Errorf("closing writer: %v", err)
+						return
+					}
+				}
+			default:
+				errCh <- fmt.Errorf("invalid request")
+				return
+			}
+		}
+	}()
+
+	var changed bool
+	sctx := util.NewClonedContext(ctx)
+	saveWithErr := func(err error) error {
```

Comment:
This function is _always_ when the server exists, even if there was an error. The `changed` flag informs this method whether or not the bucket needs saving. Note that we have to clone the server context in order to enable this logic in cases where the context is cancelled.